File: clients/python/acteon_client/a2a.py
"""A2A protocol surface for the Python ActeonClient.

Two mixins — ``_A2AClientMixin`` for sync and ``_AsyncA2AClientMixin``
for async — bolt the A2A REST surface plus the one JSON-RPC-only
method onto :class:`~acteon_client.client.ActeonClient` and
:class:`~acteon_client.client.AsyncActeonClient`. Methods mirror the
Rust client at ``crates/client/src/a2a.rs`` so cross-language code
reads the same.

Wire payloads are kept as ``dict[str, Any]`` matching the A2A JSON
shapes verbatim. A2A spec evolves and the schema is JSON-native; the
explicit dataclass route the bus module uses would force a translation
layer for every field change. The factory helpers
(:func:`make_message`, :func:`make_part_text`, etc.) cover the common
construction cases without the dataclass boilerplate.

Every authenticated call sends ``A2A-Version: 1.0`` so the server's
version negotiation honours version-pinned callers. The discovery
endpoint (``GET /.well-known/agent.json``) is issued *without* the
API-key header — the A2A spec requires it to be unauthenticated.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional
from urllib.parse import quote

from .errors import ApiError, HttpError

if TYPE_CHECKING:
    import httpx

# ---------------------------------------------------------------------
# Wire constants
# ---------------------------------------------------------------------

#: A2A protocol version this client speaks. Matches
#: ``A2A_PROTOCOL_VERSION`` in the Rust client and the server.
A2A_PROTOCOL_VERSION = "1.0"

_A2A_VERSION_HEADER = "A2A-Version"
_A2A_HEADERS = {_A2A_VERSION_HEADER: A2A_PROTOCOL_VERSION}


def _seg(s: str) -> str:
    """Percent-encode a path segment opaquely (no ``/`` passthrough).

    Mirrors ``_seg`` in ``bus.py``: a tenant or task id that happens
    to contain a slash must be escaped, not silently split into
    additional path components.
    """
    return quote(s, safe="")


def _raise_for_status(resp: "httpx.Response") -> None:
    """Translate a non-2xx response into either ``ApiError`` (with the
    server's structured error envelope) or ``HttpError`` (raw body).

    The A2A REST binding uses the same ``{"error": "..."}`` envelope
    Acteon's other handlers use, so this helper covers both A2A and
    cross-handler errors uniformly.
    """
    if 200 <= resp.status_code < 300:
        return
    try:
        data = resp.json()
        message = (
            data.get("error")
            or data.get("message")
            or f"a2a error (status {resp.status_code})"
        )
        raise ApiError(
            code=data.get("code", "A2A"),
            message=message,
            retryable=resp.status_code in (408, 425, 429) or resp.status_code >= 500,
        )
    except ValueError:
        # Not JSON — fall back to a raw HTTP error with the body text.
        raise HttpError(
            status=resp.status_code,
            message=resp.text or "a2a error",
        ) from None


# ---------------------------------------------------------------------
# Factory helpers — the common construction shapes
# ---------------------------------------------------------------------


def make_part_text(text: str) -> dict[str, Any]:
    """Build a text ``Part`` payload — the lightest A2A part shape.

    Caps: the server rejects text > 256 KiB (``MAX_PART_TEXT_BYTES``)
    at validation time.
    """
    return {"text": text}


def make_part_url(href: str) -> dict[str, Any]:
    """Build a URL-reference ``Part``. Use this for payloads that
    exceed the 256 KiB inline cap — the URL points at an external
    store the receiver fetches separately.
    """
    return {"url": href}


def make_part_data(value: Any, media_type: str = "application/json") -> dict[str, Any]:
    """Build a structured-data ``Part``. The server JSON-encodes the
    value to measure against ``MAX_PART_DATA_BYTES = 256 KiB``.
    """
    return {"data": value, "mediaType": media_type}


def make_message(
    message_id: str,
    role: str,
    parts: list[dict[str, Any]],
    *,
    task_id: Optional[str] = None,
    context_id: Optional[str] = None,
) -> dict[str, Any]:
    """Build a ``TaskMessage`` payload.

    Set ``task_id`` to thread this message into an existing Task's
    history; leave it ``None`` to mint a fresh Task on ``send_message``.

    ``role`` must be ``"user"`` or ``"agent"`` — the server validates.
    """
    msg: dict[str, Any] = {
        "messageId": message_id,
        "role": role,
        "parts": parts,
    }
    if task_id is not None:
        msg["taskId"] = task_id
    if context_id is not None:
        msg["contextId"] = context_id
    return msg


def make_push_config(
    url: str,
    *,
    id: Optional[str] = None,  # noqa: A002 - mirrors the spec field
    token: Optional[str] = None,
    authentication: Optional[dict[str, Any]] = None,
) -> dict[str, Any]:
    """Build a ``PushNotificationConfig`` body.

    ``url`` must be ``http://`` or ``https://`` — the server denies
    other schemes at registration time. ``token`` (if set) is sent
    as ``Authorization: Bearer <token>`` on every push POST and is
    treated as a secret server-side.
    """
    body: dict[str, Any] = {"url": url}
    if id is not None:
        body["id"] = id
    if token is not None:
        body["token"] = token
    if authentication is not None:
        body["authentication"] = authentication
    return body


# ---------------------------------------------------------------------
# Sync mixin
# ---------------------------------------------------------------------


class _A2AClientMixin:
    """Mixin providing the A2A protocol surface (sync)."""

    # The mixin doesn't define ``__init__``; these attributes are set
    # by the concrete client class. Stubbed for type-checkers.
    if TYPE_CHECKING:
        def _request(  # noqa: D401
            self,
            method: str,
            path: str,
            *,
            json: Optional[dict] = None,
            params: Optional[dict] = None,
            extra_headers: Optional[dict[str, str]] = None,
            skip_auth: bool = False,
        ) -> "httpx.Response": ...

    # ---- Task lifecycle ----

    def a2a_send_message(
        self,
        namespace: str,
        tenant: str,
        message: dict[str, Any],
    ) -> dict[str, Any]:
        """``POST /a2a/{namespace}/{tenant}/v1/message:send``.

        Start a new A2A Task or continue an existing one. Set
        ``message["taskId"]`` to thread into an existing Task's
        history.
        """
        resp = self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/message:send",
            json={"message": message},
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_get_task(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> dict[str, Any]:
        """``GET /a2a/{namespace}/{tenant}/v1/tasks/{id}`` — read a
        Task by id. Raises ``ApiError`` with HTTP 404 when the task
        does not exist for the caller.
        """
        resp = self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_cancel_task(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> dict[str, Any]:
        """``POST /a2a/{namespace}/{tenant}/v1/tasks/{id}:cancel``.

        The ``:cancel`` verb is part of the URL (spec §11) — the
        server splits it off in-handler. Raises ``ApiError`` with
        HTTP 409 ``TaskNotCancelable`` when the task is terminal.
        """
        resp = self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}:cancel",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    # ---- Push-notification configs ----

    def a2a_set_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config: dict[str, Any],
    ) -> dict[str, Any]:
        """``POST .../v1/tasks/{id}/pushNotificationConfigs`` — register
        or upsert a push-notification webhook for a Task.

        Use :func:`make_push_config` to build ``config``.
        """
        resp = self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}"
            f"/v1/tasks/{_seg(task_id)}/pushNotificationConfigs",
            json=config,
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_list_push_configs(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> list[dict[str, Any]]:
        """``GET .../v1/tasks/{id}/pushNotificationConfigs`` — list
        every config registered for the task.
        """
        resp = self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}"
            f"/v1/tasks/{_seg(task_id)}/pushNotificationConfigs",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_get_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config_id: str,
    ) -> dict[str, Any]:
        """``GET …/pushNotificationConfigs/{cfgId}`` — read one
        config.
        """
        resp = self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}"
            f"/pushNotificationConfigs/{_seg(config_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_delete_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config_id: str,
    ) -> None:
        """``DELETE …/pushNotificationConfigs/{cfgId}``. Raises
        ``ApiError`` with HTTP 404 when the config doesn't exist —
        the server never silently no-ops.
        """
        resp = self._request(
            "DELETE",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}"
            f"/pushNotificationConfigs/{_seg(config_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)

    # ---- Discovery ----

    def a2a_discover_agent(
        self,
        namespace: str,
        tenant: str,
    ) -> dict[str, Any]:
        """``GET /a2a/{namespace}/{tenant}/.well-known/agent.json`` —
        unauthenticated discovery endpoint.

        Sent **without** the API-key header per the A2A spec. Returns
        the tenant's ``AgentCard`` (single-card verbatim or
        aggregated across registered agents). Raises ``ApiError``
        with HTTP 404 when no agent has published a card.
        """
        resp = self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/.well-known/agent.json",
            skip_auth=True,
        )
        _raise_for_status(resp)
        return resp.json()

    def a2a_get_authenticated_extended_card(
        self,
        namespace: str,
        tenant: str,
    ) -> dict[str, Any]:
        """JSON-RPC ``agent/getAuthenticatedExtendedCard`` — the
        authenticated discovery variant.

        Issued through the JSON-RPC envelope against
        ``POST /a2a/{ns}/{tenant}`` (the A2A spec defines no REST
        counterpart for this method). The returned card has
        ``capabilities.extendedAgentCard = True``.
        """
        envelope = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "agent/getAuthenticatedExtendedCard",
        }
        resp = self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}",
            json=envelope,
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return _unwrap_jsonrpc(resp.json())


# ---------------------------------------------------------------------
# Async mixin (mirrors the sync one method-for-method)
# ---------------------------------------------------------------------


class _AsyncA2AClientMixin:
    """Mixin providing the A2A protocol surface (async)."""

    if TYPE_CHECKING:
        async def _request(  # noqa: D401
            self,
            method: str,
            path: str,
            *,
            json: Optional[dict] = None,
            params: Optional[dict] = None,
            extra_headers: Optional[dict[str, str]] = None,
            skip_auth: bool = False,
        ) -> "httpx.Response": ...

    async def a2a_send_message(
        self,
        namespace: str,
        tenant: str,
        message: dict[str, Any],
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_send_message`."""
        resp = await self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/message:send",
            json={"message": message},
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_get_task(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_get_task`."""
        resp = await self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_cancel_task(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_cancel_task`."""
        resp = await self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}:cancel",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_set_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config: dict[str, Any],
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_set_push_config`."""
        resp = await self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}"
            f"/v1/tasks/{_seg(task_id)}/pushNotificationConfigs",
            json=config,
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_list_push_configs(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
    ) -> list[dict[str, Any]]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_list_push_configs`."""
        resp = await self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}"
            f"/v1/tasks/{_seg(task_id)}/pushNotificationConfigs",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_get_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config_id: str,
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_get_push_config`."""
        resp = await self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}"
            f"/pushNotificationConfigs/{_seg(config_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_delete_push_config(
        self,
        namespace: str,
        tenant: str,
        task_id: str,
        config_id: str,
    ) -> None:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_delete_push_config`."""
        resp = await self._request(
            "DELETE",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/v1/tasks/{_seg(task_id)}"
            f"/pushNotificationConfigs/{_seg(config_id)}",
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)

    async def a2a_discover_agent(
        self,
        namespace: str,
        tenant: str,
    ) -> dict[str, Any]:
        """Async counterpart of :meth:`_A2AClientMixin.a2a_discover_agent`."""
        resp = await self._request(
            "GET",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}/.well-known/agent.json",
            skip_auth=True,
        )
        _raise_for_status(resp)
        return resp.json()

    async def a2a_get_authenticated_extended_card(
        self,
        namespace: str,
        tenant: str,
    ) -> dict[str, Any]:
        """Async counterpart of
        :meth:`_A2AClientMixin.a2a_get_authenticated_extended_card`.
        """
        envelope = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "agent/getAuthenticatedExtendedCard",
        }
        resp = await self._request(
            "POST",
            f"/a2a/{_seg(namespace)}/{_seg(tenant)}",
            json=envelope,
            extra_headers=_A2A_HEADERS,
        )
        _raise_for_status(resp)
        return _unwrap_jsonrpc(resp.json())


# ---------------------------------------------------------------------
# JSON-RPC envelope helper (shared by sync + async paths)
# ---------------------------------------------------------------------


def _unwrap_jsonrpc(body: dict[str, Any]) -> dict[str, Any]:
    """Unwrap a JSON-RPC 2.0 reply envelope. Raises ``ApiError`` when
    the envelope carries an ``error`` member.
    """
    if "error" in body and body["error"] is not None:
        err = body["error"]
        raise ApiError(
            code=str(err.get("code", "JSONRPC")),
            message=err.get("message", "JSON-RPC error"),
            retryable=False,
        )
    result = body.get("result")
    if result is None:
        raise ApiError(
            code="JSONRPC",
            message="JSON-RPC reply had neither result nor error",
            retryable=False,
        )
    return result

File: crates/client/src/a2a.rs
//! A2A protocol client helpers.
//!
//! Wraps the REST surface of the A2A endpoints exposed under
//! `/a2a/{namespace}/{tenant}/…`. The JSON-RPC transport is not
//! exposed here — the REST binding (A2A spec §11) is one-to-one with
//! the JSON-RPC methods and is easier to call directly, so callers
//! that don't need the JSON-RPC envelope can skip it entirely.
//!
//! All methods set the `A2A-Version: 1.0` header so a version-pinned
//! caller is honoured by the server. Acteon's API-key /
//! `Authorization: Bearer` header is added through the shared
//! `add_auth` helper used by every other client module.
//!
//! Re-exports the core `Task`, `TaskMessage`, `Artifact`, etc., from
//! `acteon_core` so callers don't need a direct dependency on it.

use acteon_core::{Task, TaskMessage, TaskPushNotificationConfig};

use crate::dispatch::ErrorResponse;
use crate::{ActeonClient, Error};

/// `A2A-Version` header carried on every request. Mirrors
/// `acteon_server::api::a2a::A2A_PROTOCOL_VERSION`.
pub const A2A_PROTOCOL_VERSION: &str = "1.0";

const A2A_VERSION_HEADER: &str = "A2A-Version";

/// Re-exported core types so callers can construct messages /
/// configs without depending on `acteon_core` directly.
pub use acteon_core::{
    AgentCard, Artifact, PauseKind, PushAuthentication, TaskArtifactUpdateEvent, TaskPart,
    TaskRole, TaskState, TaskStatus,
};

/// Map a non-success HTTP response to an `Error` by best-effort JSON
/// decode of the error body. Mirrors `crate::bus::map_error`.
async fn map_error(resp: reqwest::Response) -> Error {
    let status = resp.status().as_u16();
    let err = resp.json::<ErrorResponse>().await.ok();
    Error::Http {
        status,
        message: err.map_or_else(|| "a2a API error".to_string(), |e| e.message),
    }
}

impl ActeonClient {
    // -------------------------------------------------------------
    // Task lifecycle
    // -------------------------------------------------------------

    /// `POST /a2a/{namespace}/{tenant}/v1/message:send` — start a new
    /// A2A Task or continue an existing one.
    ///
    /// Set `message.task_id` to the id of an existing task to thread
    /// the message into its history; leave it `None` to mint a fresh
    /// Task.
    pub async fn a2a_send_message(
        &self,
        namespace: &str,
        tenant: &str,
        message: &TaskMessage,
    ) -> Result<Task, Error> {
        let url = format!("{}/a2a/{namespace}/{tenant}/v1/message:send", self.base_url);
        let body = MessageSendBody { message };
        let resp = self
            .add_auth(self.client.post(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .json(&body)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<Task>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// `GET /a2a/{namespace}/{tenant}/v1/tasks/{id}` — read a Task
    /// by id. Returns `Err(Error::Server { status: 404, .. })` when
    /// the task does not exist for the caller.
    pub async fn a2a_get_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
    ) -> Result<Task, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.get(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<Task>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// `POST /a2a/{namespace}/{tenant}/v1/tasks/{id}:cancel` —
    /// cancel a non-terminal Task.
    ///
    /// The `:cancel` verb suffix is part of the URL (spec §11) — the
    /// server splits it off in-handler. Returns the updated Task on
    /// success; surfaces `TaskNotCancelable` (HTTP 409) when the task
    /// is already terminal.
    pub async fn a2a_cancel_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
    ) -> Result<Task, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}:cancel",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.post(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<Task>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    // -------------------------------------------------------------
    // Push-notification configs
    // -------------------------------------------------------------

    /// `POST /a2a/{ns}/{tenant}/v1/tasks/{id}/pushNotificationConfigs`
    /// — register (or upsert, if `id` is set) a push-notification
    /// webhook for a Task. Returns the saved row with timestamps
    /// stamped.
    pub async fn a2a_set_push_config(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        body: &PushConfigInput,
    ) -> Result<TaskPushNotificationConfig, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}/pushNotificationConfigs",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.post(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .json(body)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<TaskPushNotificationConfig>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// `GET /a2a/{ns}/{tenant}/v1/tasks/{id}/pushNotificationConfigs`
    /// — list every config registered for the task. Empty Vec on
    /// existing-but-unconfigured tasks.
    pub async fn a2a_list_push_configs(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
    ) -> Result<Vec<TaskPushNotificationConfig>, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}/pushNotificationConfigs",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.get(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<Vec<TaskPushNotificationConfig>>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// `GET …/pushNotificationConfigs/{cfgId}` — read one config.
    pub async fn a2a_get_push_config(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        config_id: &str,
    ) -> Result<TaskPushNotificationConfig, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}/pushNotificationConfigs/{config_id}",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.get(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<TaskPushNotificationConfig>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// `DELETE …/pushNotificationConfigs/{cfgId}` — remove one
    /// config. `Ok(())` on success; `Err` with HTTP 404 when the
    /// config doesn't exist (the server never silently no-ops).
    pub async fn a2a_delete_push_config(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        config_id: &str,
    ) -> Result<(), Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/v1/tasks/{task_id}/pushNotificationConfigs/{config_id}",
            self.base_url
        );
        let resp = self
            .add_auth(self.client.delete(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            Ok(())
        } else {
            Err(map_error(resp).await)
        }
    }

    // -------------------------------------------------------------
    // Discovery
    // -------------------------------------------------------------

    /// `GET /a2a/{namespace}/{tenant}/.well-known/agent.json` —
    /// unauthenticated discovery endpoint. Returns the tenant's
    /// `AgentCard` (single-card verbatim or aggregated across agents).
    ///
    /// Per A2A spec this endpoint is unauthenticated, so the request
    /// is sent **without** Acteon's API-key header. A 404 surfaces
    /// as `Err(Error::Server { status: 404, .. })`.
    pub async fn a2a_discover_agent(
        &self,
        namespace: &str,
        tenant: &str,
    ) -> Result<AgentCard, Error> {
        let url = format!(
            "{}/a2a/{namespace}/{tenant}/.well-known/agent.json",
            self.base_url
        );
        let resp = self
            .client
            .get(&url)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if resp.status().is_success() {
            resp.json::<AgentCard>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(map_error(resp).await)
        }
    }

    /// JSON-RPC `agent/getAuthenticatedExtendedCard` — authenticated
    /// extended discovery card. Issued through the JSON-RPC envelope
    /// against `POST /a2a/{ns}/{tenant}` rather than a dedicated
    /// REST route (A2A spec defines no REST counterpart for this
    /// method). The returned card has
    /// `capabilities.extendedAgentCard = true`.
    pub async fn a2a_get_authenticated_extended_card(
        &self,
        namespace: &str,
        tenant: &str,
    ) -> Result<AgentCard, Error> {
        let url = format!("{}/a2a/{namespace}/{tenant}", self.base_url);
        let envelope = serde_json::json!({
            "jsonrpc": "2.0",
            "id": 1,
            "method": "agent/getAuthenticatedExtendedCard",
        });
        let resp = self
            .add_auth(self.client.post(&url))
            .header(A2A_VERSION_HEADER, A2A_PROTOCOL_VERSION)
            .json(&envelope)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;
        if !resp.status().is_success() {
            return Err(map_error(resp).await);
        }
        let env: JsonRpcReply<AgentCard> = resp
            .json()
            .await
            .map_err(|e| Error::Deserialization(e.to_string()))?;
        match (env.result, env.error) {
            (Some(card), None) => Ok(card),
            (_, Some(err)) => Err(Error::Api {
                code: err.code.to_string(),
                message: err.message,
                retryable: false,
            }),
            _ => Err(Error::Deserialization(
                "JSON-RPC reply had neither result nor error".to_string(),
            )),
        }
    }
}

// ---------------------------------------------------------------------
// Wire shapes
// ---------------------------------------------------------------------

/// Body of `POST /v1/message:send`. The server's request shape is
/// `{ "message": <TaskMessage> }`; mirrored here as a borrowed wrapper
/// so we don't have to clone the message just to serialize it.
#[derive(serde::Serialize)]
struct MessageSendBody<'a> {
    message: &'a TaskMessage,
}

/// Body of `POST .../v1/tasks/{id}/pushNotificationConfigs`. Matches
/// the JSON shape the server's `SetPushConfigInput` deserializes.
#[derive(Debug, Clone, Default, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PushConfigInput {
    /// Optional pre-allocated config id. Omit to mint a fresh one
    /// server-side.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    /// Destination URL (must be `http://` or `https://`).
    pub url: String,
    /// Optional bearer token sent in `Authorization: Bearer …`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
    /// Optional richer authentication metadata.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authentication: Option<PushAuthentication>,
}

/// Minimal JSON-RPC 2.0 response envelope used by
/// `a2a_get_authenticated_extended_card`. `T` is `Option`-wrapped on
/// purpose so `serde` doesn't require `T: Default` for the default
/// `None` branch.
#[derive(Debug, serde::Deserialize)]
struct JsonRpcReply<T> {
    result: Option<T>,
    #[serde(default)]
    error: Option<JsonRpcError>,
}

#[derive(Debug, serde::Deserialize)]
struct JsonRpcError {
    code: i64,
    message: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn message_send_body_serializes_camelcase_message() {
        // The server's wire format for message is camelCase
        // (`messageId`, `taskId`, etc.). Verify the wrapper doesn't
        // double-wrap or lose the field shape.
        let msg = TaskMessage::text("msg-1".to_string(), TaskRole::User, "hi");
        let body = MessageSendBody { message: &msg };
        let json = serde_json::to_value(&body).unwrap();
        assert!(json.get("message").is_some());
        let inner = &json["message"];
        assert_eq!(inner["messageId"], "msg-1");
        // Role enum serializes as snake_case "user" / "agent".
        assert_eq!(inner["role"], "user");
    }

    #[test]
    fn push_config_input_skips_optional_none_fields() {
        let input = PushConfigInput {
            id: None,
            url: "https://x.example.com/hook".to_string(),
            token: None,
            authentication: None,
        };
        let json = serde_json::to_value(&input).unwrap();
        assert_eq!(json["url"], "https://x.example.com/hook");
        assert!(json.get("id").is_none());
        assert!(json.get("token").is_none());
        assert!(json.get("authentication").is_none());
    }
}
